lib/network_ssl/network_ssl.c

```diff
@@ -14,6 +14,27 @@
 
 #include "network_ssl.h"
 
+#ifndef OPENSSL_VERSION_NUMBER
+#error "OPENSSL_VERSION_NUMBER must be defined"
+#endif
+
+/*
+ * LibreSSL claims to be OpenSSL 2.0, but (currently) has APIs compatible with
+ * OpenSSL 1.0.1g.
+ */
+#ifdef LIBRESSL_VERSION_NUMBER
+#undef OPENSSL_VERSION_NUMBER
+#define OPENSSL_VERSION_NUMBER 0x1000107fL
+#endif
+
+/* Compatibility for OpenSSL pre-1.1.1, found in network_ssl_compat.c */
+#if OPENSSL_VERSION_NUMBER < 0x10101000L
+int ssl_compat_write_ex(SSL *, const void *, size_t, size_t *);
```

Comment:
I debated whether to put the forward definitions before or after the `#define`.

Reply:
Neither.  They should be in `network_ssl_compat.h`.  (Which also solves the problem of forward declarations in the `.c` file.)

Reply:
ok.  That opens up another possibility: I could move the

```
#if OPENSSL_VERSION_NUMBER < 0x10101000L
#define SSL_read_ex ssl_compat_read_ex
#define SSL_write_ex ssl_compat_write_ex
#endif
```

into `network_ssl_compat.h`.  Then the entire compatibility layer could be hidden from `network_ssl.c` -- I mean, the only diff would be `#include "network_ssl_compat.h"` (and adding `network_ssl_compat.c` to `Makefile.BSD`).

Is that what you had in mind, or would you rather keep the function redefinitions in `network_ssl.c` so it's more clear what's happening?

Reply:
I think keeping the `#define`s in `network_ssl.c` is good, to make it clearer.

One option here would be to have `#if NETWORK_SSL_COMPAT_READ_EX` and similar.  Probably not useful if all we're dealing with is OpenSSL version numbers, but it might help to clean things up if we want to deal with all the different forks adding new APIs at different times.

Reply:
do you want these at the function granularity, or a logical grouping?  i.e. `#if NETWORK_SSL_COMPAT_READ_WRITE_EX`, since I can't imagine an API which only had one of `_read_` and `_write_`.

Reply:
Logical groupings would be fine.

Reply:
ooo, that would also allow us to avoid compiling `_compat_` functions which relied on deprecated functions!  Definitely a plus!